### PR TITLE
Support Clang 8 and add clang4 and clang8 as Travis targets

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Work around a compiler bug in clang8
- Add clang4 and clang8 as Travis targets.

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

File: containers/Dockerfile.buildenv [hunk 1]
```diff
@@ -72,46 +76,59 @@ RUN apt-get update -y \
...
-    && spack compiler find \
-    && sed -i 's@fc: null@fc: /usr/bin/gfortran@' \
-    /root/.spack/linux/compilers.yaml \
-    && sed -i 's@f77: null@f77: /usr/bin/gfortran@' \
-    /root/.spack/linux/compilers.yaml
-RUN /work/spack/bin/spack install cmake \
-    && /work/spack/bin/spack install --no-checksum catch@2.1.0 \
-    && /work/spack/bin/spack install brigand@master \
-    && /work/spack/bin/spack install blaze \
-    && /work/spack/bin/spack install gsl%gcc@6.5.0 \
```
Comment (moxcodes) [inline containers/Dockerfile.buildenv:107]: For my own clarification: Was this previously redundant, or was there something about versions that needed this? It looks like we already had it in the apt.

File: .travis.yml [hunk 2]
```diff
@@ -140,18 +140,42 @@ jobs:
...
+    dist: trusty
+    env: BUILD_TYPE=Release COMPILER=clang-5.0
+    name: clang-5 release
+  - &Clang8Debug
+    os: linux
+    compiler: clang-8
+    dist: trusty
+    env: BUILD_TYPE=Debug COMPILER=clang-8
+    name: clang-8 debug
+  - &Clang8Release
```
Comment (fmahebert) [inline .travis.yml:173]: For my understanding: why clang versions 4,5,8 but not 6,7 ?

File: .travis/BuildLinux.sh [hunk 3]
```diff
@@ -61,18 +49,20 @@ if [ -z "${RUN_CLANG_TIDY}" ] \
...
         time make test-libs-elliptic -j2
         time make test-libs-evolution -j2
-        time make test-libs-numerical-algorithms -j2
     fi
 
     if [[ ${TRAVIS_BUILD_STAGE_NAME} = "Build test libraries" ]]; then
-        time make test-libs-data-structures -j2
+        time make test-libs-numerical-algorithms -j2
+        # Build DataStructures in serial because the DataVector tests
+        # are very memory intensive
```
Comment (fmahebert) [inline .travis/BuildLinux.sh]: I suggest clarifying: "very memory intensive to compile"

File: .travis/BuildLinux.sh [hunk 4]
```diff
@@ -61,18 +49,20 @@ if [ -z "${RUN_CLANG_TIDY}" ] \
...
+        time make test-libs-data-structures -j1
         time make test-libs-pointwise-functions -j2
-        time make test-libs-other -j2
```
Comment (fmahebert) [inline .travis/BuildLinux.sh:70]: Do I correctly understand that with this line being moved to a different Travis build stage, the "Build test libraries" stage no longer builds all test libraries? If so, consider renaming to something more accurate.

File: src/NumericalAlgorithms/Spectral/Spectral.hpp [hunk 5]
```diff
@@ -72,23 +72,31 @@ std::ostream& operator<<(std::ostream& os,
                          const Quadrature& quadrature) noexcept;
 /// \endcond
 
+namespace detail {
+constexpr size_t minimum_number_of_points(
+    const Basis basis, const Quadrature quadrature) noexcept {
+  // Silence compiler warnings.
+  (void)basis;
+  (void)quadrature;
```
Comment (fmahebert) [inline src/NumericalAlgorithms/Spectral/Spectral.hpp]: I can see how `(void)basis` might silence a warning, but why is `(void)quadrature` needed? The value is used in the if statement below...

File: src/NumericalAlgorithms/Spectral/Spectral.hpp [hunk 6]
```diff
@@ -72,23 +72,31 @@ std::ostream& operator<<(std::ostream& os,
...
+  if (quadrature == Quadrature::Gauss) {
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Spectral/Spectral.hpp:78]: Another clarification question: The previous version appeared to produce a compile-time value for the appropriate basis. If my understanding of `if`s with `constexpr` is correct, this no longer produces a compile-time value if the basis is valid; is that right?
However, it seems it didn't generate any errors in Travis, so either I'm wrong or we never used them at compile-time to begin with.

File: containers/Dockerfile.buildenv [hunk 7]
```diff
@@ -72,46 +76,59 @@ RUN apt-get update -y \
...
-    && /work/spack/bin/spack install libsharp -openmp -mpi \
-    && /work/spack/bin/spack install libxsmm%gcc@6.5.0 \
-    && /work/spack/bin/spack install yaml-cpp@develop%gcc@6.5.0 \
-    && /work/spack/bin/spack install benchmark%gcc@6.5.0
+# We install dependencies not available through apt manually rather than using
+# Spack since Spack ends up building a lot of dependencies from scratch
+# that we don't need. Thus, not building the deps with Spack reduces total
+# build time of the Docker image.
```
Comment (fmahebert) [inline containers/Dockerfile.buildenv:76]: 👍 for this comment. Relatedly: now that we aren't using Spack in the container, do we still need to install LMod (new line 70ish, not part of this PR)?

**Review by moxcodes — DISMISSED** [review]
Some very minor clarification questions, I'm happy with the changes.

**Review by fmahebert — COMMENTED** [review]
I'm overall happy with this, but do have a few questions. If any changes result, please squash them immediately.

File: containers/Dockerfile.buildenv (hunk 1 above)
Comment (nilsdeppe) [reply containers/Dockerfile.buildenv:107]: If memory serves me correctly then way back when this started (two years ago) the LTS ubuntu was 14.04 and so it didn't have any newer versions of these packages, or even things like catch at all. Spack was suggested by Erik Schnetter (and generally works well), but what it does is it compiles e.g. CMake from scratch for every compiler version unless you specify really long modifier strings for each package giving each dependency and exactly what compiler that dependency should be built with. While good for HPC systems, that seemed like overkill here. Compiling CMake takes something like 10 minutes, and compiling it 2-3 times meant this process took a very long time. 

If you're referring to `gsl` specifically, I think this was merely an oversight when the apt installation was added (presumably my fault :) )

File: src/NumericalAlgorithms/Spectral/Spectral.hpp (hunk 6 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/Spectral.hpp:78]: Very good question! C++14 relaxes what can be done inside constexpr functions. With C++14 you can do if statements and for loops (don't remember while loops off the top of my head). You are also allowed to write long functions, whereas in C++11 you were only allowed a single return statement with ternaries. I'm sure you can imagine how gross some functions looked back then :) Because of the increases in constexpr, some people argue every function should be marked constexpr "just in case". With C++20 `new` and `delete` will work at compile time, and so will `std::vector` and `std::string`, which means a tremendous amount of code can be constexpr then. Of course, whether it actually gets evaluated at compile time is a different story. But it will remove more and more of the template metaprogramming in favor of regular programming where the code runs at compile time. The C++ committee is hoping to completely remove the need for template metaprogramming (not templates themselves, though) by around C++26 or C++29. Does that explain things?

File: .travis.yml (hunk 2 above)
Comment (nilsdeppe) [reply .travis.yml:173]: To give a compromise between the number of targets and how long it takes to build. There's no technical reason, but clang4 is the oldest we can support, clang5 was used for IWYU, and clang8 since something broke in going from 7 to 8. Does that make sense?

File: .travis/BuildLinux.sh (hunk 3 above)
Comment (nilsdeppe) [reply .travis/BuildLinux.sh]: I can't (or at least super not easily) profile the usage on Travis so I can't give numbers, but I'm happy to add more to the comment if you give me some guidelines as to what to say :)

File: .travis/BuildLinux.sh (hunk 4 above)
Comment (nilsdeppe) [reply .travis/BuildLinux.sh:70]: Yes. @kidder since you came up with the names and these stages, what are your thoughts?

File: src/NumericalAlgorithms/Spectral/Spectral.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/Spectral.hpp]: It does not. Once the other comments have been settled I'll remove.

File: containers/Dockerfile.buildenv (hunk 7 above)
Comment (nilsdeppe) [reply containers/Dockerfile.buildenv:76]: Yes, I was partly too lazy to wait for the image to build again :P I'll remove once the other comments are settled :)

File: .travis.yml (hunk 2 above)
Comment (fmahebert) [reply .travis.yml:173]: Yep! I assumed it was something like this. Thanks for clarifying.

File: .travis/BuildLinux.sh (hunk 3 above)
Comment (fmahebert) [reply .travis/BuildLinux.sh]: Oh, sorry I wasn't super clear, I meant clarifying "memory intensive" -> "memory intensive to compile" in your comment, so as to avoid any ambiguity about compile-time vs run-time memory use. :)

File: src/NumericalAlgorithms/Spectral/Spectral.hpp (hunk 6 above)
Comment (moxcodes) [reply src/NumericalAlgorithms/Spectral/Spectral.hpp:78]: Excellent, that nicely clarifies for me, thanks.

File: .travis.yml (hunk 2 above)
Comment (nilsdeppe) [reply .travis.yml]: Ah, good point! Will do :)

File: .travis/BuildLinux.sh (hunk 4 above)
Comment (fmahebert) [reply .travis/BuildLinux.sh:70]: "Build most expensive test libraries" is cumbersome but hopefully descriptive?

File: .travis/BuildLinux.sh (hunk 3 above)
Comment (nilsdeppe) [reply .travis/BuildLinux.sh]: Ah, good point! Will do :)

File: .travis/BuildLinux.sh (hunk 4 above)
Comment (nilsdeppe) [reply .travis/BuildLinux.sh:70]: I'm sort of in favor of naming the stages things like `Build stage 1`, `Build stage 2`, `Build stage 3, run tests, clang-tidy, doxygen` but :man_shrugging:

Comment (kidder) [reply .travis/BuildLinux.sh:70]: yeah since the test libraries are now built in all 3 stages, and we'll likely need to rebalance again in the future, `Build stage N` is fine with me...

Comment (fmahebert) [reply .travis/BuildLinux.sh:70]: Okay with me as well

**Comment by nilsdeppe** [comment]
- Forced push squashed commits (please check I didn't mess anything up :) )
- Pushed updated docker container that no longer contains LMod

**Review by moxcodes — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]
